core/src/main/java/org/apache/iceberg/MetricsConfig.java

```diff
@@ -104,38 +108,40 @@ public static MetricsConfig forTable(Table table) {
    * @param table an Iceberg table
    */
   public static MetricsConfig forPositionDelete(Table table) {
-    MetricsConfig config = new MetricsConfig();
+    ImmutableMap.Builder<String, MetricsMode> columnModes = ImmutableMap.builder();
 
-    config.columnModes.put(MetadataColumns.DELETE_FILE_PATH.name(), MetricsModes.Full.get());
-    config.columnModes.put(MetadataColumns.DELETE_FILE_POS.name(), MetricsModes.Full.get());
+    columnModes.put(MetadataColumns.DELETE_FILE_PATH.name(), MetricsModes.Full.get());
+    columnModes.put(MetadataColumns.DELETE_FILE_POS.name(), MetricsModes.Full.get());
 
     MetricsConfig tableConfig = forTable(table);
 
-    config.defaultMode = tableConfig.defaultMode;
+    MetricsMode defaultMode = tableConfig.defaultMode;
     tableConfig.columnModes.forEach((columnAlias, mode) -> {
       String positionDeleteColumnAlias = DOT.join(MetadataColumns.DELETE_FILE_ROW_FIELD_NAME, columnAlias);
-      config.columnModes.put(positionDeleteColumnAlias, mode);
+      columnModes.put(positionDeleteColumnAlias, mode);
     });
 
-    return config;
+    return new MetricsConfig(columnModes.build(), defaultMode);
   }
 
   private static MetricsConfig from(Map<String, String> props, SortOrder order) {
-    MetricsConfig spec = new MetricsConfig();
+    Map<String, MetricsMode> columnModes = new HashMap<>();
+    MetricsMode defaultMode;
     String defaultModeAsString = props.getOrDefault(DEFAULT_WRITE_METRICS_MODE, DEFAULT_WRITE_METRICS_MODE_DEFAULT);
     try {
-      spec.defaultMode = MetricsModes.fromString(defaultModeAsString);
+      defaultMode = MetricsModes.fromString(defaultModeAsString);
     } catch (IllegalArgumentException err) {
       // Mode was invalid, log the error and use the default
       LOG.warn("Ignoring invalid default metrics mode: {}", defaultModeAsString, err);
-      spec.defaultMode = MetricsModes.fromString(DEFAULT_WRITE_METRICS_MODE_DEFAULT);
+      defaultMode = MetricsModes.fromString(DEFAULT_WRITE_METRICS_MODE_DEFAULT);
     }
 
     // First set sorted column with sorted column default (can be overridden by user)
-    MetricsMode sortedColDefaultMode = sortedColumnDefaultMode(spec.defaultMode);
+    MetricsMode sortedColDefaultMode = sortedColumnDefaultMode(defaultMode);
     Set<String> sortedCols = SortOrderUtil.orderPreservingSortedColumns(order);
-    sortedCols.forEach(sc -> spec.columnModes.put(sc, sortedColDefaultMode));
+    sortedCols.forEach(sc -> columnModes.put(sc, sortedColDefaultMode));
 
+    MetricsMode defaultModeFinal = defaultMode;
```

Comment:
I guess overall it is not so clean, but given we already set the defaultMode before and below, we can just continue using the same variable instead of introducing a new one?

Reply:
I think it's because the object passed to lambda must not be mutated.

Reply:
I thought it just has to be effectively final, this makes it a bit confusing below where we use "defaultModeFinal" and "defaultMode". I think this is fine if the compiler wants it but let's use the same variable in both spots below
```java
mode = defaultModeFinal
```
and
```java
return new MetricsConfig(columnModes, defaultModeFinal);
```

Reply:
> I thought it just has to be effectively final

correct. and seems it needs to be assigned once, and here it's assigned in an `if`.

> `return new MetricsConfig(columnModes, defaultModeFinal);`

the variable `defaultModeFinal` exists for lambda only. it's not meant to change logic flow, or supersede the existing variable, once assigned. that's why i'd use it only for the lambda and "forget" later.